src/game/client/baseheightmap.cpp

```diff
@@ -968,89 +968,100 @@ void BaseHeightMapRenderObjClass::Do_The_Light(VertexFormatXYZDUV2 *vb,
+                                          + (bottom_right_height - bottom_left_height) * fract_x)
+                    * HEIGHTMAP_SCALE;
             }
 
+            // Calculate the normal vector if pos is provided
             if (pos != nullptr) {
-                int i4 = x_extent * (i2 - 1) + i1;
-                int i5 = x_extent * i2 + i1;
-                int i6 = x_extent + i5;
-                int i7 = x_extent * (i2 + 2) + i1;
-                unsigned char c1 = data[i5];
-                unsigned char c2 = data[i5 + 1];
-                unsigned char c3 = data[x_extent + 1 + i5];
-                unsigned char c4 = data[x_extent + i5];
-                unsigned char c5 = data[i4];
-                unsigned char c6 = data[i4 + 1];
-                unsigned char c7 = data[i7 + 1];
-                unsigned char c8 = data[i7];
-                float f11 = c2 - data[i5 - 1];
-                float f12 = data[i5 + 2] - c1;
-                float f13 = data[x_extent + 2 + i5] - c4;
-                float f14 = c4 - c5;
-                float f15 = c3 - c6;
-                float f16 = c7 - c2;
-                float f17 = c8 - c1;
-                float f18 = (1.0f - f5) * f11 + f5 * f12;
-                float f19 = (1.0f - f5) * f12 + f5 * f13;
-                float f20 = f18 * (1.0f - f6) + f6 * f19;
-                float f21 = (1.0f - f5) * f14 + f5 * f17;
-                float f22 = (1.0f - f5) * f15 + f5 * f16;
-                float f23 = f21 * (1.0f - f6) + f6 * f22;
-                Vector3 v1;
-                Vector3 v2;
-                Vector3 v3;
-                v1.Set(32.0f, 0.0f, f20);
-                v2.Set(0.0f, 32.0f, f23);
-                Vector3::Normalized_Cross_Product(v1, v2, &v3);
-                pos->x = v3.X;
-                pos->y = v3.Y;
-                pos->z = v3.Z;
+                int prev_row_cell_index = map_width * (index_y - 1) + index_x;
+                int current_cell_index = map_width * index_y + index_x;
+                int next_row_cell_index = map_width + current_cell_index;
+                int next_next_row_cell_index = map_width * (index_y + 2) + index_x;
+                unsigned char current_cell_height_data = height_data[current_cell_index];
+                unsigned char right_neighbor_height_data = height_data[current_cell_index + 1];
+                unsigned char bottom_right_neighbor_height_data = height_data[map_width + 1 + current_cell_index];
+                unsigned char bottom_neighbor_height_data = height_data[map_width + current_cell_index];
+                unsigned char top_neighbor_height_data = height_data[prev_row_cell_index];
+                unsigned char top_right_neighbor_height_data = height_data[prev_row_cell_index + 1];
+                unsigned char down_two_right_neighbor_height_data = height_data[next_next_row_cell_index + 1];
+                unsigned char down_two_neighbor_height_data = height_data[next_next_row_cell_index];

```

Comment:
Yes, I was under the assumption that 0,0 was top left. As I understand it, map_width * index_y + index_x is the index for (x, y). Rather than using a 2D array, we think of map_width as one row worth of items, so y*number_of_rows gets us which row we're in, and +/- x moves us around within a given row. I'll rewrite it so 0,0 is bottom left.

Reply:
Done